### PR TITLE
Added "select" statement

### PR DESCRIPTION
**Pull request by dwb**
This adds a new statement, "select". Before this addition it was
not possible to, say, chuck multiple events to `now` and resume on
the first event signalled, or to implement a timeout on waiting on
an event. Examples for these are in the also-added tests. Inspiration
for this feature comes of course from [the Go language](https://golang.org/ref/spec#Select_statements).

I believe I have also cleaned up some bits around event reference
counting, but I appreciate this is a bit of a twisty area and I could
well have missed something.

C++ isn't my usual language; I've tried to follow conventions and
not go off on one, but sorry if I've misapprehended a couple of things.
I did take the liberty of moving to C++14; I hope this is agreeable. I
didn't see that this project had any particularly stringent compiler
compatibility requirements, but happy to be corrected on that.

Also sorry for just plopping this here without discussing it first. I
completely understand if you think this is the wrong direction for some
reason; I just had an idea and felt like writing it. If it needs major re-work
then I won't be offended!

I can write a lot more of course if it would be helpful, but I think the tests
are fairly clear. I've had this sitting around for a few months after getting it
working and been putting off writing it up "properly", but I figured it was
better to get it up and start the conversation at least.

### CODE REVIEW

**Comment by markcerqueira** [comment]
To get our new GitHub Actions checks to run and pass could you please rebase this off of `main` and force push to your branch / PR? Thanks!

**Comment by gewang** [comment]
Greetings! first of all — thank you for your time and energy in creating this PR! 

**TL;DR**
We recently did a "semi-cataclysmic" cleanup of this repository and as a result, all PRs created before June 15, 2023 are referencing invalid content. We are closing all existing PRs for now to avoid confusion. If you would, please do a fresh `git clone` of the repository, and manually apply the changes against the latest, and resubmit the PR. Sorry for any confusion (and the delay) and thank you for your continued contribution to ChucK!

**Detailed Explanation**
The Great ChucK Repo Cleanup of 2023 has been completed! As a result, the overall chuck repo size has been reduced from 721MB to 25MB -- since git works by cloning the entire repo and all history, this massively improves the speed of `git clone`, `git submodule update --init` in repos that reference chuck, and the size of each cloned repo on disk.

To accomplish this, we had to "rewrite the git history," which is a semi-cataclysmic process for git repos. As a result, all existing clones of the chuck, miniAudicle, and chunity repos as well as branches and outstanding pull-requests that reference the older repo _**are now invalid**_. Please do a fresh clone of the `chuck` / `miniAudicle` / `chunity` repos and copy any important changes to these fresh clones before making any further changes to avoid headaches!

NOTE: do not use 'git pull' with your existing, pre-cleanup chuck branches and forks, as it will seem to work BUT will put your local repo into a messed up state. A fresh `git clone` is necessary before proceeding with chuck development! We believe that the cleaner, smoother ChucK dev experience will be worth this short-term inconvenience.

If you have any difficulty porting any existing branches/PRs to the new repo, please let us know and we will do our best to help! (The Chuck Community Discord can be a good place for this: https://discord.gg/ENr3nurrx8; also feel free to continue commenting on this closed PR)

Apologies for any confusion that may arise, and **thank you** for being involved in ChucK development!!
